File: 백준/Gold/15989. 1， 2， 3 더하기 4/1， 2， 3 더하기 4.py
import sys

input = sys.stdin.read().strip().split('\n')

t = int(input[0])
testcase = list(map(int, input[1:]))
maxValue = max(testcase)

d = [[0] * 4 for i in range(maxValue + 1)]

d[1][1] = 1
d[2][1] = 1
d[2][2] = 1
d[3][1] = 1
d[3][2] = 1
d[3][3] = 1

for i in range(4, maxValue + 1):
    d[i][1] = d[i - 1][1]
    d[i][2] = d[i - 2][1] + d[i - 2][2]
    d[i][3] = d[i - 3][1] + d[i - 3][2] + d[i - 3][3]

for test in testcase:
    print(d[test][1] + d[test][2] + d[test][3])

File: 백준/Gold/15989. 1， 2， 3 더하기 4/1， 2， 3 더하기 4.js
const fs = require('fs');
const input = fs.readFileSync(0, 'utf8').trim().split('\n');

const t = parseInt(input[0]);
const testcase = input.slice(1).map(Number);
const maxValue = Math.max(...testcase);

const dp = Array.from({ length: maxValue + 1 }, () => Array(4).fill(0));

dp[1][1] = 1; 
dp[2][1] = 1;
dp[2][2] = 1;
dp[3][1] = 1;
dp[3][2] = 1;
dp[3][3] = 1;

for (let i = 4; i <= maxValue; i++) {
    dp[i][1] = dp[i - 1][1];
    dp[i][2] = dp[i - 2][1] + dp[i - 2][2]; 
    dp[i][3] = dp[i - 3][1] + dp[i - 3][2] + dp[i - 3][3];
}

for (let test of testcase) {
    console.log(dp[test][1] + dp[test][2] + dp[test][3]); 
}
